### PR TITLE
remove error info from the engine release for fireball/issues/6721

### PR DESCRIPTION
**Pull request by knoxHuang**
Re: cocos-creator/fireball#6721

Changelog:
 * 从引擎发布版中剔除异常报错信息

> Makes sure these boxes are checked before submitting your PR - thank you!
>
- [x] If your pull request has gone "stale", you should **rebase** your work on top of the latest version of the upstream branch.
- [x] If your commit history is full of small, unimportant commits (such as "fix pep8" or "update tests"), **squash** your commits down to a few, or one, discreet changesets before submitting a pull request.
- To official teams:
  - [x] Check that your javascript is following our [style guide](https://github.com/cocos-creator/fireball/blob/dev/.github/CONTRIBUTING.md) and end files with a newline
  - [x] Document new code with comments in source code based on [API Docs](https://github.com/cocos-creator/fireball#api-docs)
  - [x] Make sure any runtime log information in `cc.log` , `cc.error`, `cc.warn` or `cc.assert` has been moved into `DebugInfos.js` with an ID

### CODE REVIEW

File: DebugInfos.js [hunk 1]
```diff
@@ -24,9 +24,20 @@ if (CC_DEBUG) {
...
+        "1025": "cc.Sequence.initWithTwoActions(): arguments must all be non nil", //cc.Sequence.initWithTwoActions
+        "1026": "cc.RepeatForever.initWithAction(): action must be non null", //cc.RepeatForever.initWithAction
+        "1027": "cc.Spawn.initWithTwoActions(): arguments must all be non null", //cc.Spawn.initWithAction
+        "1028": "cc.ReverseTime.initWithAction(): action must be non null", //cc.ReverseTime.initWithAction
+        "1029": "cc.ReverseTime.initWithAction(): the action was already passed in.", //cc.ReverseTime.initWithAction
+        "1030": "cc.Animate.initWithAnimation(): animation must be non-NULL", //cc.Animate.initWithAction
         //configuration: 1100
         "1100": "Expected 'data' dict, but not found. Config file: %s", //loadConfigFile
         "1101": "Please load the resource first : %s", //loadConfigFile_2
+        "1102": "Please load the resource first : %s", // loadConfigFile
```
Comment (jareguo) [inline DebugInfos.js]: 1102 和 1101 重复

File: DebugInfos.js [hunk 2]
```diff
@@ -571,6 +613,23 @@ if (CC_DEBUG) {
...
+        "7228": "_ccsg.TMXLayer.getTileGIDAt(): pos should be non-null",//CCSGTMXLayer getTileGIDAt
+        "7229": "_ccsg.TMXLayer.getTileGIDAt(): invalid position",//CCSGTMXLayer getTileGIDAt
+        "7230": "_ccsg.TMXLayer.setTileGID(): pos should be non-null",//CCSGTMXLayer setTileGID
+        "7231": "_ccsg.TMXLayer.setTileGID(): invalid position",//CCSGTMXLayer setTileGID
+        "7232": "_ccsg.TMXLayer.getTileFlagsAt(): pos should be non-null",//CCSGTMXLayer getTileFlagsAt
+        "7233": "_ccsg.TMXLayer.getTileFlagsAt(): invalid position",//CCSGTMXLayer getTileFlagsAt
+        "7234": "_ccsg.TMXLayer.removeTileAt(): pos should be non-null",//CCSGTMXLayer removeTileAt
+        "7235": "_ccsg.TMXLayer.removeTileAt(): invalid position",//CCSGTMXLayer removeTileAt
+        "7236": "_ccsg.TMXLayer.getTileAt(): pos should be non-null",//CCSGTMXLayer getTileAt
+        "7237": "_ccsg.TMXLayer.getTileAt(): pos should be non-null",//CCSGTMXLayer getTileAt
```
Comment (jareguo) [inline DebugInfos.js]: 上面两条重复了

File: DebugInfos.js [hunk 3]
```diff
@@ -633,6 +704,27 @@ if (CC_DEBUG) {
...
+    var errorMapUrl = 'https://github.com/cocos-creator/engine/blob/master/EngineErrorMap.md';
+
+    cc._LogInfos.format = function (id) {
+        var type = 'error';
+        if (arguments.length === 1) {
+            return CC_DEBUG ? cc._LogInfos[id] : type + ' ' + id + ', please go to ' + errorMapUrl + '#' + id + ' to see details.'
+        }
+        if (CC_DEBUG) {
+            let argsArr = cc.js.shiftArguments.apply(null, arguments);
+            return cc._LogInfos[id].concat(argsArr);
```
Comment (jareguo) [inline DebugInfos.js]: 别忘了 cc.js.formatStr 啊操

**Review by jareguo — CHANGES_REQUESTED** [review]

**Comment by jareguo** [comment]
这个新的 format 接口未能和旧接口很好的关联在一起，而且定义两遍 errorMapUrl 你能忍？

建议定义两个新的 _LogInfos 接口

```js
function raw (id, type) {
    return CC_DEBUG ? cc._LogInfos[id] : `${type} ${id}, please go to ${errorMapUrl}#${id} to see details.`;   // 引擎可以用 template string
}

function format (msg, argsArray) {
    return CC_DEBUG ? cc.js.formatStr.apply(null, [msg].concat(argsArray)) : msg + ' Arguments: ' + argsArray.join(', ');
}
```

原先 genLogFunc 内部实现就可以简化为

```js
        'use strict';   // 不可省略
        var msg = cc._LogInfos.raw(id, type);
        if (arguments.length === 1) {
            func(msg);
        }
        else if (arguments.length === 2) {    // 新增分支，用于优化性能
            func(CC_DEBUG ? cc.js.formatStr(msg, arguments[1]) : msg + ' Arguments: ' + arguments[1]);
        }
        else {
            func(cc._LogInfos.format(msg, cc.js.shiftArguments.apply(null, arguments)));
        }
```

所以原来调用 cc._LogInfos.format 的地方要改成调用 cc._LogInfos.e (这个名字来自 `catch (e)`)，e 的实现只需要依赖 format

```
function e () {
        var msg = raw(arguments[0], 'ERROR');   // type 大写，方便和 errorID 区分
        if (arguments.length === 1) {
            return msg;
        }
        else {
            return format(msg, cc.js.shiftArguments.apply(null, arguments)); 
        }
}
```

这样虽然代码量可能省不了多少，但是可读性会好一些。

**Comment by jareguo** [comment]
上面的方法我没测试过，你测看看

File: DebugInfos.js [hunk 4]
```diff
@@ -633,6 +700,28 @@ if (CC_DEBUG) {
         "8400": "Wrong type arguments, 'filePath' must be a String.", // cc.audioEngine.play
     };
     cc._LogInfos = logs;
+
+    var errorMapUrl = 'https://github.com/cocos-creator/engine/blob/master/EngineErrorMap.md';
+
+    function raw (id, type) {
+        return false ? cc._LogInfos[id] : `${type} ${id}, please go to ${errorMapUrl}#${id} to see details.`;
```
Comment (jareguo) [inline DebugInfos.js]: false？

Comment (knoxHuang) [reply DebugInfos.js]: 无视，我提错了

**Review by jareguo — APPROVED** [review]

**Comment by knoxHuang** [comment]
好了

File: cocos2d/core/components/CCRendererInSG.js [hunk 5]
```diff
@@ -96,7 +96,7 @@ var RendererInSG = cc.Class({
 
     _replaceSgNode: function (sgNode) {
         if (CC_EDITOR && !(sgNode instanceof _ccsg.Node)) {
-            throw new Error("Invalid sgNode. It must be an instance of _ccsg.Node");
+            throw new Error(cc._LogInfos.e(1636));
```
Comment (jareguo) [inline cocos2d/core/components/CCRendererInSG.js]: 麻烦把这个报错去掉，这个是调试用的，本来就不会进入到非 CC_DEBUG 版本

**Review by jareguo — CHANGES_REQUESTED** [review]

**Review by jareguo — APPROVED** [review]

**Pull request merged**